crates/core/executor/src/memory.rs

```diff
@@ -0,0 +1,290 @@
+use std::mem::{replace, size_of};
+
+use serde::{Deserialize, Serialize};
+use vec_map::VecMap;
+
+use crate::events::MemoryRecord;
+
+/// The base 2 logarithm of the (maximum) page size in bytes.
+pub const LOG_PAGE_SIZE: usize = 12;
+/// The base 2 logarithm of the length of each page, considered as an array of `Option<MemoryRecord>`.
+pub const LOG_PAGE_LEN: usize =
+    LOG_PAGE_SIZE - size_of::<Option<MemoryRecord>>().next_power_of_two().ilog2() as usize;
```

Comment:
I commented on this above in the PR. A couple concerns I had:
- Writing a bunch of `const` functions is extra work.
- Rust has many restrictions on const generics, e.g. type generic parameters cannot be used in const generic positions. By genericizing all the constants, we may restrict ourselves. This should not currently be an issue, but it is an issue if, say, we wanted to replace the `Page` `VecMap`s with arrays.

I feel that it's best to use logarithm when dealing with powers of two. This enables the use of the shift operators, which can be more efficient than dividing. I don't believe the compiler will "optimize" division/multiplication by powers of two because of truncation/overflow issues (at least for multiplication).

For that specific case in `indices`, it may be more efficient to do division and remainder together (since x86 uses a single instruction to do both), rather than shifting and masking. However, these sort of bitwise operations seem to be rather common in performance-critical code, which is why I use them.

In summary, I don't know which approach is more efficient. I would likely have to write micro-benchmarks to detect differences.

Reply:
Moved all the constants into the `impl` block.